cpp/include/cudf/types.h

```diff
@@ -85,6 +85,17 @@ typedef struct {
 	// here we can also hold info for decimal datatype or any other datatype that requires additional information
 } gdf_dtype_extra_info;
 
+
+/**---------------------------------------------------------------------------*
+ * @brief A struct to hold a scalar (single) value and its type information
+ * 
+ *---------------------------------------------------------------------------**/
+typedef struct {
+  void*  data;      /**< Pointer to the scalar data */
```

Comment:
This should not be a `void*`. That's unavoidable, and somewhat reasonable, for a `gdf_column`, but here it should just be a union of the different possible column element types.

Using a pointer means:

* A `gdf_scalar` can _never_ be owning.
* Probably-unrecoverable loss of optimization opportunities in tight loops.
* Inability to compare two `gdf_scalar`s trivially.
* Confusing to the users/callers: When, if ever, can they free the memory to which the `gdf_scalar` is pointing to?
* Since pointers cannot typically be used both on the device and on the host, this us in the on-host-vs-on-device mess, which is bad enough for columns. A proper plain-old-data scalar would avoid this altogether
* Additional risk of null/invalid pointer dereferencing. While it's true we can't avoid these risks with `gdf_column` - we should not introduce more of them if we can help it.
* 

Finally,:

* It isn't really consistent with `gdf_column`. This has an extra level of indirection which `gdf_column` does not; and it is is either non-nullable, or nullable in a different way than `gdf_column`. 
* Since a pointer is typically 8 bytes, there's no space savings from not having the union.
* If we _did_ want something consistent with `gdf_column` for scalars, we'd use... well, we'd use `gdf_column`s constrained to have length 1.

Bottom line:  Vehemently oppose the design of this structure. Instead, use a distinguished union and a boolean, e.g. like so:
```
typedef struct {
    union {
        cudf::column_element_type_t< GDF_INT8      > int8;
        cudf::column_element_type_t< GDF_INT16     > int16;
        cudf::column_element_type_t< GDF_INT32     > int32;
        cudf::column_element_type_t< GDF_INT64     > int64;
        cudf::column_element_type_t< GDF_FLOAT32   > float32;
        cudf::column_element_type_t< GDF_FLOAT64   > float64;
        cudf::column_element_type_t< GDF_DATE32    > date32;
        cudf::column_element_type_t< GDF_DATE64    > date64;
        cudf::column_element_type_t< GDF_TIMESTAMP > timestamp;
        cudf::column_element_type_t< GDF_CATEGORY  > category;
    } datum;
    gdf_dtype dtype;
    bool valid; 
} gdf_scalar;
```
or whatever @jrhemstad 's dtype-to-actual-type trait is called.

FYI @harrism , @williamBlazing , @jrhemstad .

Reply:
@eyalroz I understand most of your argument, but you claim that scalars *must* be nullable without any argument for why you feel that to be true. 

What is an example use case where an invalid scalar must be passable to a binary op?

Reply:
@harrism : The motivation is that (nullable) columns can hold the NULL value. Their element type is T disjoint-union Unit. NULL is "just" another value; it has some special semantics, but I could just as well may pass NULL to a binary op as I may pass a non-null value. In particular, my scalar may be the result of some lookup operation in a nullable column. There's no reason my code should special-case the lookup result being NULL rather than 123 or what-not.

Also, a binary operator applied to a column and a scalar can (and should be) considered equivalent to using a uniform column of the same value as an operand. And, obviously, uniform columns of a certain element type can, in particular, be uniformly NULL.

Reply:
The reason I do not support a union is because it is incompatible with the `type_dispatcher`.

The `type_dispatcher` is predicated on the idea that for a given `gdf_dtype`, it dispatches you the correct `T`, and you can `static_cast` your type-erased date (`void*`) to `T`.

However, you cannot cast a `union` to one of it's member types:

https://wandbox.org/permlink/zTfeMxDUtBTNiNuL

Furthermore, using a `union` increases the amount of code that needs to be manually updated whenever a new `gdf_dtype` is added.

Lastly,

```
 cudf::column_element_type_t< GDF_INT8      > int8;
```
No such trait exists. In fact, when designing the `type_dispatcher` and I proposed such a trait to map `gdf_dtype`s to `T`s, you _assured_ me that there was no reason for such a trait to exist :) 

You're thinking of the inverse of mapping a `T` to a `gdf_dtype`. https://github.com/rapidsai/cudf/pull/952/files#diff-7a5c6cb1fc7d9f8ada8ffb4e15d130efR170

Reply:
Okay, so you _could_ make a `union` work like this:

```
union U {
    int myInt;
    float myFloat;
};

int main() { 
    U foo;
    foo.myInt = 42;
    std::cout << (int) *reinterpret_cast<int*>(&foo) << std::endl;
}
```

Reply:
I think we're confusing C++ and C here. For the C API, there is obviously no type dispatcher and we have to work with what C offers. For C++, of course we wouldn't use a union - we'd use a variant. Not `std::variant` that's not available to us, but there are nice variant implementations available, like [Michael Park's.](https://github.com/mpark/variant).

Reply:
> I think we're confusing C++ and C here. For the C API, there is obviously no type dispatcher and we have to work with what C offers. For C++, of course we wouldn't use a union - we'd use a variant. Not `std::variant` that's not available to us, but there are nice variant implementations available, like [Michael Park's.](https://github.com/mpark/variant).

Switching to C++ internals that could support a `variant` is months away and vastly outside the scope of this PR. As such, it is mandatory that a `gdf_scalar` be compatible with the `type_dispatcher`. 

A `union` will work via the example I provided above. As such, I support using a `union` for the `gdf_scalar` (sorry Devavret :( ).

Reply:
> Switching to C++ internals that could support a variant is months away

You don't have to switch _everything_ to support variants. Right now we don't have any code which takes a `gdf_scalar`. So instead of gradually switching to supporting a struct-with-void-* scalar we gradually switch to an `mp::variant` scalar.

But - union is certainly a reasonable initial choice.

Reply:
Yeah, your union-based struct is good.

We'd either need to create the ` cudf::column_element_type_t` trait you suggested, or just manually specify the types.

Reply:
@jrhemstad : Didn't you already merge a PR with type <-> enum value conversion mechanisms?

Reply:
> No such trait exists. In fact, when designing the type_dispatcher and I proposed such a trait to map gdf_dtypes to Ts, you assured me that there was no reason for such a trait to exist :)

I don't remember what I said, but you didn't need it for the type dispatcher. Also, that trait could be implemented using a tuple type and a decltype + std::get + declval, I think.